clients/include/testing_dense2csx.hpp

```diff
@@ -0,0 +1,396 @@
+    //
+    int  MN          = (dirA == HIPSPARSE_DIRECTION_ROW) ? M : N;
+    auto m_dense_val = hipsparse_unique_ptr{device_malloc(sizeof(T) * LD * N), device_free};
+    auto nnzPerRowColumn_managed
+        = hipsparse_unique_ptr{device_malloc(sizeof(int) * MN), device_free};
+    auto nnzTotalDevHostPtr_managed
+        = hipsparse_unique_ptr{device_malloc(sizeof(int) * 1), device_free};
+
+    T*   d_dense_val          = (T*)m_dense_val.get();
+    int* d_nnzPerRowColumn    = (int*)nnzPerRowColumn_managed.get();
+    int* d_nnzTotalDevHostPtr = (int*)nnzTotalDevHostPtr_managed.get();
+    if(!d_nnzPerRowColumn || !d_nnzTotalDevHostPtr || !d_dense_val)
+    {
+        verify_hipsparse_status_success(HIPSPARSE_STATUS_ALLOC_FAILED,
+                                        "!h_nnzPerRowColumn || !d_nnzPerRowColumn || !d_dense_val");
+        return HIPSPARSE_STATUS_ALLOC_FAILED;
+    }
+
+    //
+    // Initialize the entire allocated memory.
+    //
+    for(int i = 0; i < LD; ++i)
+    {
+        for(int j = 0; j < N; ++j)
+        {
+            h_dense_val[j * LD + i] = T(-1.0);
+        }
+    }
+
+    //
+    // Initialize a random dense matrix.
+    //
+    srand(0);
+    gen_dense_random_sparsity_pattern(M, N, h_dense_val.data(), LD, 0.2);
+
+    //
+    // Transfer.
+    //
+    CHECK_HIP_ERROR(
+        hipMemcpy(d_dense_val, h_dense_val.data(), sizeof(T) * LD * N, hipMemcpyHostToDevice));
+
+    int nnz;
+    CHECK_HIPSPARSE_ERROR(hipsparseSetPointerMode(handle, HIPSPARSE_POINTER_MODE_HOST));
+    CHECK_HIPSPARSE_ERROR(hipsparseXnnz(
+        handle, DIRA, M, N, descr, (const T*)d_dense_val, LD, d_nnzPerRowColumn, &nnz));
+
+    //
+    // Transfer.
+    //
+    CHECK_HIP_ERROR(hipMemcpy(
+        h_nnzPerRowColumn.data(), d_nnzPerRowColumn, sizeof(int) * DIMDIR, hipMemcpyDeviceToHost));
+
+    auto m_csx_row_col_ptr
+        = hipsparse_unique_ptr{device_malloc(sizeof(int) * (DIMDIR + 1)), device_free};
+    auto m_csx_val         = hipsparse_unique_ptr{device_malloc(sizeof(T) * nnz), device_free};
+    auto m_csx_col_row_ind = hipsparse_unique_ptr{device_malloc(sizeof(int) * nnz), device_free};
+
+    int* d_csx_row_col_ptr = (int*)m_csx_row_col_ptr.get();
+    int* d_csx_col_row_ind = (int*)m_csx_col_row_ind.get();
+    T*   d_csx_val         = (T*)m_csx_val.get();
+
+    if(!d_csx_row_col_ptr || !d_csx_val || !d_csx_col_row_ind)
+    {
+        CHECK_HIP_ERROR(hipErrorOutOfMemory);
+        return HIPSPARSE_STATUS_ALLOC_FAILED;
+    }
+
+    std::vector<int> cpu_csx_row_col_ptr(DIMDIR + 1);
+    std::vector<T>   cpu_csx_val(nnz);
+    std::vector<int> cpu_csx_col_row_ind(nnz);
+    if(!cpu_csx_row_col_ptr.data() || !cpu_csx_val.data() || !cpu_csx_col_row_ind.data())
+    {
+        CHECK_HIP_ERROR(hipErrorOutOfMemory);
+        return HIPSPARSE_STATUS_ALLOC_FAILED;
+    }
+
+    if(argus.unit_check)
+    {
+
+        //
+        // Compute the reference host first.
+        //
+
+        host_dense2csx<DIRA, T>(M,
```

Comment:
You write this so that it would return a status but you are not checking it here. Did you want to check it?

Reply:
;)